### ISSUE TITLE
Performance comparison between onnxruntime-web and onnxruntime-python

### ISSUE THREAD

**Issue by jobergum** [microsoft/onnxruntime#9161]
First, thank you for all the work on the new onnxruntime web! It looks awesome with a lot of great examples. Super excited about this!

TL;DR the onnxruntime-web is approx 25x slower than the python onnxruntime inference with the same input sequence.
Does this match up with your testing? 

I'm curious about the inference performance. I did a quick test building on the minimal web example using the script tag and compared the performance with similar intra threads and the onnxruntime python binding using a model that I'm familiar with which can be downloaded from https://data.vespa.oath.cloud/onnx_models/ms-marco-MiniLM-L-6-v2-quantized.onnx. 

For the same input sequence I get these numbers with intrathreads 1:

- onnxruntime-web 100 ms
- onnxruntime-python 4 ms

Also, using intrathreads on on onnxruntime-web does not seem to have any impact either.

*onnxruntime-web*
```
<html>
    <header>
        <title>ONNX Runtime JavaScript examples: Quick Start</title>
    </header>
    <body>
        <!-- import ONNXRuntime Web from CDN -->
        <script src="https://cdn.jsdelivr.net/npm/onnxruntime-web/dist/ort.min.js"></script>
        <script>
            async function main() {
                try {
                    const options = {executionProviders: ['wasm'],intraOpNumThreads: 1}
                    const session = await ort.InferenceSession.create('./ms-marco-MiniLM-L-6-v2-quantized.onnx', options);
                    const input =   BigInt64Array.from([BigInt(101), BigInt(2349), BigInt(102), BigInt(2349), BigInt(102)]);
                    const mask =    BigInt64Array.from([BigInt(1), BigInt(1),BigInt(1), BigInt(1), BigInt(1)]);
                    const type =    BigInt64Array.from([BigInt(0), BigInt(0),BigInt(0),BigInt(1), BigInt(1)]);
                    
                    const input_ids = new ort.Tensor('int64', input, [1,5]);
                    const attention_mask = new ort.Tensor('int64', mask, [1,5]);
                    const token_type_ids = new ort.Tensor('int64', type, [1,5]);
                    
                    const feeds = { input_ids: input_ids, token_type_ids: token_type_ids, attention_mask:attention_mask};
                    const start = Date.now();
                    const results = await session.run(feeds,['output_0'], {logSeverityLevel:3});
                    const duration = Date.now() - start;
                    console.log("MiniLM inference score " + results['output_0'].data[0] + " took " + duration + "ms");

                } catch (e) {
                    document.write(`failed to inference ONNX model: ${e}.`);
                }
            }

            main();
        </script>
    </body>
</html>
```
The above takes around 100 ms in Chrome while the following using python3 on the same node takes 4 ms. 

*onnxruntime-python* 
```
import onnxruntime as ort
session = ort.InferenceSession("ms-marco-MiniLM-L-6-v2-quantized.onnx")
import numpy as np
from time import time

options = ort.SessionOptions()
options.intra_op_num_threads = 1
session = ort.InferenceSession("ms-marco-MiniLM-L-6-v2-quantized.onnx", options)

input = { 'input_ids': np.array([[101, 2349, 102, 2349, 102]]), 'attention_mask': np.array([[1,1,1,1,1]]), 'token_type_ids':np.array([[0,0,0,1,1]])}
start = int(time() * 1000)   
score, = session.run(input_feed=input,output_names=['output_0'])[0][0]
duration = int(time()*1000) - start
print("MiniLM inference using onnxruntime py {} ms, score={}".format(duration,score))
```

**Comment by fs-eire** [comment]
Thank you for your feedback.

onnxruntime-python is a python language binding of the native ONNX Runtime, which utilize all available platform and OS features to accelerate ONNX Runtime model inferencing. This is similar to ORT Java binding, C# binding and Node.js (javascript) binding.

onnxruntime-web is not using the native ONNX Runtime. Because of the restriction on browser, a website cannot run native libraries on client's device. We utilize WebAssembly to make ONNX Runtime running inside a browser. This means it will not going to beat native performance. We do expect onnxruntime-web is slower than native ORT. If you want to use JavaScript on server, onnxruntime-node (ORT node.js binding) is recommended.

However, 25x slower is too slow, it is not expected. I will take a look at the model and profile to see how we can improve.

**Comment by jobergum** [comment]
Thank you @fs-eire for getting back to me with a long reply, deeply appreciate it!

I'm not expecting it to be close to the native onnx-rt with all the magic hw accelerations through MLAS etc but was hoping for guidance on if the observations matched up with your expectations. 

With native onnx-rt using MiniLM/transformer based models we can reduce latency by tuning intrathreads to a sweet spot between latency/throughput, but in my testing with onnxruntime-web I was not seeing any latency improvement. 

Thanks again!

**Comment by fs-eire** [comment]
With some analysis, I figured out that the top 2 slowest Nodes are **DynamicQuantizeMatMul** and **MatMulIntegerToFloat**. They both call [`MlasGemmBatch`](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/mlas/lib/qgemm.cpp#L145) in **qgemm**.

It looks like MLAS does not enable SIMD for QGEMM on WebAssembly. This should be a big perf improvement to support U8x16 (WebAssembly uses SIMD128).

**Comment by jobergum** [comment]
Thanks for the work on this @fs-eire ! Looking forward to seeing the improvement from https://github.com/microsoft/onnxruntime/pull/9191 once merged and in a release.

**Comment by fs-eire** [comment]
The change is merged. please try the latest master build ( use [this link](https://dev.azure.com/onnxruntime/onnxruntime/_build/results?buildId=504679&view=results) for build artifacts ).

By the way, it takes some time for VM to JIT wasm for the first run. You may give a few warmup runs before check the performance number.

**Comment by jobergum** [comment]
Awesome @fs-eire ! I'm happy to report a very significant improvement. **6X**. 🚀 🚀 🚀  

I've expanded the example I gave to handle text input and for the same input length I see 6X latency improvement after downloading the artifacts. Before (released) and after (build artifacts) inference latency for the exact same input. 

![image](https://user-images.githubusercontent.com/20928528/135582423-e436cba0-76a9-422b-8ab3-56c126af6334.png)

![image](https://user-images.githubusercontent.com/20928528/135582463-9b69bd3a-09a7-4e83-a846-23f1de932542.png)

Thanks a lot for optimizing this, onnx-web is awesome!
